### DISCUSSION TITLE
Task Scheduling very slow

### DISCUSSION THREAD

**Discussion by itsai220** [concourse/concourse#6860] [Help & Support]
Hi posting here if anyone can help me
ever since I've upgraded concourse from 5.8.x to 6.4.1, I see task scheduling extremely slow.
I happen to see this here that this task initialized but didn't get to spawning and attached till 3 minutes later.
There are a lot of things happening in between, like checking resource, sweeping volumes etc. 

> {"timestamp":"2021-04-09T21:50:34.195628473Z","level":"info","source":"atc","message":"atc.tracker.notify.run.task-step.initializing","data":{"build":77747,"job":"appk8s-tests","job-id":5334,"pipeline":"pc-integrations","session":"23.41356.6.25","step-name":"create-placeholder-platform-env"}}
{"timestamp":"2021-04-09T21:53:08.544825135Z","level":"info","source":"atc","message":"atc.tracker.notify.run.task-step.spawning","data":{"build":77747,"job":"appk8s-tests","job-id":5334,"pipeline":"pc-integrations","session":"23.41356.6.25","step-name":"create-placeholder-platform-env"}}
{"timestamp":"2021-04-09T21:53:13.441581698Z","level":"info","source":"atc","message":"atc.tracker.notify.run.task-step.attached","data":{"build":77747,"job":"appk8s-tests","job-id":5334,"pipeline":"pc-integrations","session":"23.41356.6.25","step-name":"create-placeholder-platform-env"}}

My question here is is there any configuration I need to look out for to speed up task initialization, spawning, attach so tasks are scheduled and running faster ?

I currently have 
buildTrackerInterval: 10s
LidarScannerInterval defaults to 10s
Does resource check too often slows down build track ?

Thanks

**Comment by taylorsilva** [comment]
Most common cause for slow task start-up is volume streaming. Volume streaming happens when the task lands on a worker that doesn't have all the volumes required to satisfy the `inputs` of a task. The selected `image` can also be part of this, depending on your image setup for the task.

What container placement strategy are you using? What baggageclaim driver are you using?

**Reply by itsai220** [reply]
Hi Taylor,
thanks for the reply

container placement strategy is limit-active-tasks
baggage claim driver is set to naive

**Reply by taylorsilva** [reply]
Cool, so two things:

1. `limit-active-tasks` makes it very likely that every task in your jobs (if you have multiple tasks per jobs, with prior tasks outputting artifacts the following jobs need) is landing on a different worker. This means artifact streaming is happening between each task.
2. The `naive` baggageclaim driver is not very performant. Every time a step needs an artifact generated from another step, the entire contents of the artifact are copied. If you have a lot of artifacts and they are large then this can take a while. This would be an issue even if all your tasks land on the same worker; artifacts won't be streamed anywhere, saving your time, but you'll lose time as the artifacts are copied on-disk. You should try to use the `overlay` driver instead as its much more performant.

**Reply by itsai220** [reply]
Thanks Taylor,

I'll try setting baggageclaim driver to overlay and see if there's an improvement in performance.

We had baggageclaim and limit-active-task set since 3.8.x and I've only seen performance (task scheduling) slowness when i upgraded to 6.4.1, was there something changed that would impact how artifacts straming between tasks ?

**Reply by taylorsilva** [reply]
`limit-active-task` was added in v5.5.0: https://github.com/concourse/concourse/releases/tag/v5.5.0

You must have been using a different strategy before then or got your version numbers mixed up.

Have you tried using `volume-locality`? That's the default container placement strategy. Tasks will generally start up faster but you risk overloading a worker with too many tasks.

**Reply by itsai220** [reply]
ah sorry, I honestly thought i had limit-active-task a lot earlier on, felt like i had that value set for a while but for sure since upgrading to 6.4.1 it's been significantly slower at scheduling.

yes, we've tried other placement strategies before, and we decided to use limit-active-task to avoid overloading a worker when limit-active-task became available.

**Reply by itsai220** [reply]
Hi Taylor,

So an update, we've tried using volume-locality and switching baggage claim driver to overlay and the task spin up is still very slow, some tasks took 5mins to execute. Please let me know if there's anything else i can try to speed up task execution.

**Reply by taylorsilva** [reply]
Are you talking about initialization time or execution time? `initialization` is when all the volume streaming and setup happens for the step.  Execution time is when the step is running.
<img width="429" alt="image" src="https://user-images.githubusercontent.com/4583837/118153266-b6c14680-b3e3-11eb-89ca-c19fe01c900c.png">

I forget when this got added and too lazy to look it up, but do you see `selected worker: ....` as the first line of build output? Would be a nice data point because it clearly shows when Concourse picked  worker for the task.
<img width="431" alt="image" src="https://user-images.githubusercontent.com/4583837/118153597-14ee2980-b3e4-11eb-96a9-745b60b26d4c.png">

**Reply by itsai220** [reply]
cool, i didn't know the initialization time is shown on the checkmark, please see attached screenshot. This shows that it took 3 mins to initialize the task.

 ![Screen Shot 2021-05-13 at 9 55 46 AM](https://user-images.githubusercontent.com/48690477/118159232-9f795d80-b3d1-11eb-9ea5-c79e6f359528.png)

**Reply by taylorsilva** [reply]
Nice! The hard part now is figuring out what volumes are causing it to be so slow? How large are the inputs typically for that task?

**Reply by itsai220** [reply]
Thanks Taylor,
I've spotted this and started optimizing one of our pipeline and realize we had couple volumes we pass that were quite large, largest one being around 2.5gb. We always been passing this volume in that pipeline but we only noticed the slowness after we upgraded.

The above screenshot is a different pipeline which doesn't have large volume mentioned above, but we are using a generic task.yml that does list that volume as optional: true. Would listing a long list of optional volume slow things down ?

**Reply by taylorsilva** [reply]
I wouldn't think having a lot of optional inputs would slow down initialization. I could be wrong though. It's getting harder to provide further suggestions without having a reproducible or env I can mess around with :(

**Reply by itsai220** [reply]
Hey Taylor, sorry for delay reply
i can definitely show you what we have in our generic task yaml and give you an idea on the size of each volume if that would help to reproduce ?
i'm not sure what else i can provide to help debug this, and the slowness adds 15-20m to each of our builds so it is slowing things down quite a bit.

**Reply by itsai220** [reply]
Update: looks like there is an option to select streaming artifacts compression option, defaults to gzip but looks sounds like zstd is a more optimal option. trying this out and see if it cuts down the task initialization time.

**Comment by StevenArmstrong** [comment]
We had a similar issue and it was database IO and when we increased the
disk space on the AWS RDS instance it sped everything up as more disk means
more IO. Check if that is your bottleneck as it is nicely hidden.

**Reply by taylorsilva** [reply]
Interesting! I never even considered that db IO could be a bottleneck here. Thanks for sharing!

**Reply by itsai220** [reply]
Thanks for sharing!
in terms of db disk space we are only using ~10% of the db disk right now running on a 4core cpu, that hasn't changed since the upgrade
for worker disk however i have set it to 500gb on ssd, we do hit close to maxing out on busy days but the slowness happens even when there are no other builds running.
